### ISSUE TITLE
IOPS/BW/Lat Logging While Running

### ISSUE THREAD

**Issue by iwoloschin** [axboe/fio#588]
Is it possible to output the IOPS/BW/Latency logging files while an fio job is running, or are these always held in memory until the job is complete?  If it is not possible, would it be reasonable to request this as a feature?  If it's not feasible to write to disk (since that might impact a running test) would it instead be possible to output to the console, or even better yet push directly to a time series database like InfluxDB or Graphite?

**Comment by iwoloschin** [comment]
I'm using some of the options from here: http://fio.readthedocs.io/en/latest/fio_doc.html#measurements-and-reporting

Specifically, I'm interesting right now in `--write_bw_log` and `write_iops_log` right now.

**Comment by sitsofe** [comment]
@iwamatsu fio only writes those logs at the end of the run and not on the fly. The performance hit of writing such logs (e.g. for every I/O on fast devices) would be huge...

**Comment by iwoloschin** [comment]
@sitsofe, if you're writing *every* millisecond sure, but I think averages once a second, or even ten seconds, is probably sufficient for this kind of casual logging and you could still dump the full logs at the end of a test.  You also obviously shouldn't dump this to the device under test!  I think it'd even potentially be reasonable to suggest such a feature be only used on a tmpfs based filesystem.

**Comment by sitsofe** [comment]
@iwoloschin You're right that if the averaging period is large enough the impact should be mild.

At any rate it would require new development work and I think there's an argument what you're asking for is similar to that of https://github.com/axboe/fio/issues/130 (although you're asking for it via logging files).

**Comment by iwoloschin** [comment]
So I had a feeling this was new development work, that's good to know I haven't misread the documentation 😄.

I think for my purposes, I just want the data to be easily accessible, I don't think it really matters whether it's a log file, lines printed to a console, or even pushed directly to a time series database.  I think #130 would probably suffice for what I'm trying to do for now, though I think live log files or direct integration with a time series database are also useful...though maybe only to me if no one else has requested this before.

**Comment by sitsofe** [comment]
@iwoloschin OK in that case can I duplicate this against #130 (and I'd suggest that you add notification on that issue so if it is ever solved you'll know)?

**Comment by iwoloschin** [comment]
Sure, that sounds reasonable, thanks!

**Comment by sitsofe** [comment]
Duplicate of #130

**Comment by szaydel** [comment]
My knee-jerk reaction is that there are benefits to doing this, but not necessarily logging to a file or files, but rather just generating structured output to stdout or stderr. While many people may have different target destinations for such data, I think most can figure out how to create sinks when data just needs to be scraped from stdout. It is not much different than scraping a log file, arguably, but I feel like it is more universally useful.

**Comment by sitsofe** [comment]
@szaydel Ideally what would happen is that JSON `--status-interval` would be able to also work in "only values from this interval mode" and it would at least give people one option for doing live reporting...

**Comment by aggieNick02** [comment]
Ended up here a few years later investigating fio logging performance, file output, etc. Right now logging can have a significant impact on the performance observed for a disk (on a speedy SSD the difference between logging only bandwidth every 100ms, and logging all bandwidth and latency for every IO, was on the order of 100MB/s). So any logging to a file would need to be carefully done in a different thread, I'd imagine.

Further, just the time to write the log files that are held in memory can be significant. I did some experiments replacing the fprintf calls with doing the formatting manually (just printing numbers and characters, and using a buffer), and that took the time for a 60 second job to flush its 15 million log entries (times 5 logs = bw, i/o, lat, clat, slat) from 20 seconds to 5 seconds.

Logging in general is really an interesting thing to think about for fio. The per sample log files also take a long time to ingest. Obviously it would take some work, but I think doing per-sample logging would be well-served by a protobuf/protobuf-like output format that could be both written and read much more quickly, and would be significantly smaller on disk too.
